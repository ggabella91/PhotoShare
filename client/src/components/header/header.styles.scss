.header {
  background-color: #3b5ab1;
  min-height: 15vh;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  font-size: calc(10px + 2vmin);
  color: white;
  padding: 0;

  .app-name {
    justify-content: left;
    margin-left: 20px;
    margin-top: 40px;
    padding-bottom: 0;

    .title {
      font-weight: 560;
    }
  }
}
