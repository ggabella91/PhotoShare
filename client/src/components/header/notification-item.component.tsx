import React from 'react';
import { useNavigate } from 'react-router-dom';
import { Grid, Button, Typography, Avatar } from '@mui/material';
import { Notification } from '../../redux/follower/follower.types';

interface NotificationItemProps {
  notification: Notification;
}

const NotificationItem: React.FC<NotificationItemProps> = ({
  notification,
}) => {
  const { message, createdAt, postId } = notification;
  const navigate = useNavigate();

  const handleClickNotificationItem = () => {
    if (postId) {
      navigate(`/p/${postId}`);
    }
  };

  return (
    <Grid sx={{ display: 'flex', flexDirection: 'row', height: '75px' }}>
      <Grid>
        <Avatar
          src={''
            // userInfo?.profilePhotoFileString
            //   ?
            // `data:image/jpeg;base64,${
              // '' /**user profile photo's file string */
            // }`
            // : ''
          }
          alt={''}
          sx={{ height: '56px', width: '56px' }}
        />
      </Grid>
      <Button
        sx={{
          display: 'flex',
          flexDirection: 'column',
          width: '100%',
          minWidth: 'unset',
          height: '100%',
          textTransform: 'unset',
          padding: 1,
          color: 'black',
          borderBottom: '1px solid rgb(219,219,219)',
          borderRadius: 0,
          textAlign: 'start',
          '&:hover': {
            backgroundColor: 'unset',
          },
        }}
        disableRipple
        onClick={handleClickNotificationItem}
      >
        <Typography sx={{ width: 'inherit', fontSize: '14px ' }}>
          {message}
        </Typography>
        <Typography sx={{ width: 'inherit', fontSize: '12px ' }}>
          {new Date(createdAt).toDateString()}
        </Typography>
      </Button>
    </Grid>
  );
};

export default NotificationItem;
