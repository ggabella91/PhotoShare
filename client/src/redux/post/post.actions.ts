import { UserInfoAndOtherData } from '../../components/user-info/user-info.component';
import { PostModalDataToFeed } from '../../components/feed-post-container/feed-post-container.component';

import {
  Post,
  PostDataReq,
  PostFileReq,
  Reaction,
  ReactionReq,
  ArchivePostReq,
  DeleteReactionReq,
  PostFile,
  PostError,
  PostActions,
  PostActionTypes,
  PostMetaData,
} from './post.types';

export const createPostStart = (post: FormData): PostActionTypes => ({
  type: PostActions.CREATE_POST_START,
  payload: post,
});

export const createPostSuccess = (post: Post): PostActionTypes => ({
  type: PostActions.CREATE_POST_SUCCESS,
  payload: post,
});

export const createPostFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.CREATE_POST_FAILURE,
  payload: error,
});

export const createPostReactionStart = (
  reactionReq: ReactionReq
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_START,
  payload: reactionReq,
});

export const createPostReactionSuccess = (
  reaction: Reaction
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_SUCCESS,
  payload: reaction,
});

export const createPostReactionFailure = (
  error: PostError
): PostActionTypes => ({
  type: PostActions.CREATE_POST_REACTION_FAILURE,
  payload: error,
});

export const getPostReactionsStart = (postId: string): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_START,
  payload: postId,
});

export const getPostReactionsSuccess = (
  reactions: Reaction[]
): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_SUCCESS,
  payload: reactions,
});

export const getPostReactionsFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_REACTIONS_FAILURE,
  payload: error,
});

export const clearPostStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_STATUSES,
  payload: null,
});

export const updateProfilePhotoStart = (photo: FormData): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_START,
  payload: photo,
});

export const updateProfilePhotoSuccess = (photo: Post): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_SUCCESS,
  payload: photo,
});

export const updateProfilePhotoFailure = (
  error: PostError
): PostActionTypes => ({
  type: PostActions.UPDATE_PROFILE_PHOTO_FAILURE,
  payload: error,
});

export const clearProfilePhotoStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_PROFILE_PHOTO_STATUSES,
  payload: null,
});

export const getPostDataStart = (
  postDataReq: PostDataReq
): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_START,
  payload: postDataReq,
});

export const getPostDataSuccess = (postDataArray: Post[]): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_SUCCESS,
  payload: postDataArray,
});

export const addPostDataToFeedArray = (
  postDataArray: Post[]
): PostActionTypes => ({
  type: PostActions.ADD_POST_DATA_TO_FEED_ARRAY,
  payload: postDataArray,
});

export const getPostDataFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_DATA_FAILURE,
  payload: error,
});

export const getPostFileStart = (fileReq: PostFileReq): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_START,
  payload: fileReq,
});

export const getPostFileSuccess = (file: PostFile): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_SUCCESS,
  payload: file,
});

export const getPostFileFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.GET_POST_FILE_FAILURE,
  payload: error,
});

export const getProfilePhotoFileSuccess = (
  file: PostFile
): PostActionTypes => ({
  type: PostActions.GET_PROFILE_PHOTO_FILE_SUCCESS,
  payload: file,
});

export const archivePostStart = (
  archiveReq: ArchivePostReq
): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_START,
  payload: archiveReq,
});

export const archivePostSuccess = (message: string): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_SUCCESS,
  payload: message,
});

export const archivePostFailure = (error: Error): PostActionTypes => ({
  type: PostActions.ARCHIVE_POST_FAILURE,
  payload: error,
});

export const deleteReactionStart = (
  deleteReactionReq: DeleteReactionReq
): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_START,
  payload: deleteReactionReq,
});

export const deleteReactionSuccess = (message: string): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_SUCCESS,
  payload: message,
});

export const deleteReactionFailure = (error: PostError): PostActionTypes => ({
  type: PostActions.DELETE_REACTION_FAILURE,
  payload: error,
});

export const getOtherUserProfilePhotoFileSuccess = (
  file: string
): PostActionTypes => ({
  type: PostActions.GET_OTHER_USER_PROFILE_PHOTO_FILE_SUCCESS,
  payload: file,
});

export const getUserPhotoForFollowArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_FOLLOW_ARRAY_SUCCESS,
  payload: postFile,
});

export const getUserPhotoForSuggestionArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_SUGGESTION_ARRAY_SUCCESS,
  payload: postFile,
});

export const getUserPhotoForReactorArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_REACTOR_ARRAY_SUCCESS,
  payload: postFile,
});

export const clearArchivePostStatuses = (): PostActionTypes => ({
  type: PostActions.CLEAR_ARCHIVE_POST_STATUSES,
  payload: null,
});

export const clearFollowPhotoFileArray = (): PostActionTypes => ({
  type: PostActions.CLEAR_FOLLOW_PHOTO_FILE_ARRAY,
  payload: null,
});

export const clearSuggestionPhotoFileArray = (): PostActionTypes => ({
  type: PostActions.CLEAR_SUGGESTION_PHOTO_FILE_ARRAY,
  payload: null,
});

export const clearPostFilesAndData = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_FILES_AND_DATA,
  payload: null,
});

export const clearPostState = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_STATE,
  payload: null,
});

export const setCommentToDelete = (
  deleteReactionReq: DeleteReactionReq
): PostActionTypes => ({
  type: PostActions.SET_COMMENT_TO_DELETE,
  payload: deleteReactionReq,
});

export const setShowCommentOptionsModal = (
  showCommentOptionsModal: boolean
): PostActionTypes => ({
  type: PostActions.SET_SHOW_COMMENT_OPTIONS_MODAL,
  payload: showCommentOptionsModal,
});

export const clearPostReactions = (): PostActionTypes => ({
  type: PostActions.CLEAR_POST_REACTIONS,
  payload: null,
});

export const setPostMetaDataForUser = (
  postMetaData: PostMetaData
): PostActionTypes => ({
  type: PostActions.SET_POST_META_DATA_FOR_USER,
  payload: postMetaData,
});

export const setPostLikingUsersArray = (
  postLikingUsersArray: UserInfoAndOtherData[]
): PostActionTypes => ({
  type: PostActions.SET_POST_LIKING_USERS_ARRAY,
  payload: postLikingUsersArray,
});

export const setShowPostLikingUsersModal = (
  showPostLikingUsersModal: boolean
): PostActionTypes => ({
  type: PostActions.SET_SHOW_POST_LIKING_USERS_MODAL,
  payload: showPostLikingUsersModal,
});

export const setFeedPagePostModalData = (
  postModalDataToFeed: PostModalDataToFeed
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_MODAL_DATA,
  payload: postModalDataToFeed,
});

export const setFeedPagePostModalShow = (
  postModalDataShow: boolean
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_MODAL_SHOW,
  payload: postModalDataShow,
});

export const setFeedPagePostOptionsModalShow = (
  postOptionsModalDataShow: boolean
): PostActionTypes => ({
  type: PostActions.SET_FEED_PAGE_POST_OPTIONS_MODAL_SHOW,
  payload: postOptionsModalDataShow,
});

export const setClearFeedPagePostModalState = (
  clearFeedPagePostModalState: boolean
): PostActionTypes => ({
  type: PostActions.SET_CLEAR_FEED_PAGE_POST_MODAL_STATE,
  payload: clearFeedPagePostModalState,
});

// Actions for feed-post-container

export const getFeedPostFileSuccess = (file: PostFile): PostActionTypes => ({
  type: PostActions.GET_FEED_POST_FILE_SUCCESS,
  payload: file,
});

export const getFeedPostReactionsSuccess = (
  reactions: Reaction[]
): PostActionTypes => ({
  type: PostActions.GET_FEED_POST_REACTIONS_SUCCESS,
  payload: reactions,
});

export const getUserPhotoForFeedReactorArraySuccess = (
  postFile: PostFile
): PostActionTypes => ({
  type: PostActions.GET_USER_PHOTO_FOR_FEED_REACTOR_ARRAY_SUCCESS,
  payload: postFile,
});
