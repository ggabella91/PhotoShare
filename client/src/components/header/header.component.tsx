import React, { useState, useEffect } from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { createStructuredSelector } from 'reselect';
import { AppState } from '../../redux/root-reducer';
import { User } from '../../redux/user/user.types';
import { selectCurrentUser } from '../../redux/user/user.selectors';
import { signOutStart } from '../../redux/user/user.actions';
import { PostFileReq, UserType } from '../../redux/post/post.types';
import {
  selectProfilePhotoKey,
  selectProfilePhotoFile,
} from '../../redux/post/post.selectors';
import { getPostFileStart } from '../../redux/post/post.actions';

import './header.styles.scss';

interface HeaderProps {
  currentUser: User | null;
  profilePhotoKey: string | null;
  profilePhotoFile: string | null;
  getPostFileStart: typeof getPostFileStart;
  signOutStart: typeof signOutStart;
}

export const Header: React.FC<HeaderProps> = ({
  currentUser,
  profilePhotoKey,
  profilePhotoFile,
  getPostFileStart,
  signOutStart,
}) => {
  const [photoFile, setPhotoFile] = useState<string | null>(null);

  let bucket: string;

  if (process.env.NODE_ENV === 'production') {
    bucket = 'photo-share-app-profile-photos';
  } else {
    bucket = 'photo-share-app-profile-photos-dev';
  }

  useEffect(() => {
    if (profilePhotoKey) {
      getPostFileStart({
        s3Key: profilePhotoKey,
        bucket,
        user: UserType.self,
      });
    } else if (!profilePhotoFile && currentUser && currentUser.photo) {
      getPostFileStart({
        s3Key: currentUser.photo,
        bucket,
        user: UserType.self,
      });
    }
  }, [profilePhotoKey, currentUser]);

  useEffect(() => {
    if (profilePhotoFile) {
      setPhotoFile(profilePhotoFile);
    }
  }, [profilePhotoFile]);

  return (
    <div className='header'>
      <NavLink className='app-name' to='/'>
        <h1 className='title'>PhotoShare</h1>
      </NavLink>
      {currentUser ? (
        <div>
          <NavLink to={`/${currentUser.username}`} className='avatar'>
            {photoFile ? (
              <img
                className='profile-photo'
                src={`data:image/jpeg;base64,${photoFile}`}
                alt='profile-pic'
              />
            ) : null}
            {!photoFile ? (
              <div className='photo-placeholder'>
                <span className='photo-placeholder-text'>No photo</span>
              </div>
            ) : null}
          </NavLink>
          <nav className='header-menu'>
            <NavLink className='link' to='/settings'>
              Settings
            </NavLink>
            <NavLink className='link' to='/' onClick={signOutStart}>
              Sign Out
            </NavLink>
          </nav>
        </div>
      ) : null}
    </div>
  );
};

interface LinkStateProps {
  currentUser: User | null;
  profilePhotoKey: string | null;
  profilePhotoFile: string | null;
}

const mapStateToProps = createStructuredSelector<AppState, LinkStateProps>({
  currentUser: selectCurrentUser,
  profilePhotoKey: selectProfilePhotoKey,
  profilePhotoFile: selectProfilePhotoFile,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  getPostFileStart: (fileReq: PostFileReq) =>
    dispatch(getPostFileStart(fileReq)),
  signOutStart: () => dispatch(signOutStart()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
