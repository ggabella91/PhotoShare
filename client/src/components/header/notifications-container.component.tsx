import React, { useState, useEffect, useRef } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Grid, Typography } from '@mui/material';
import NotificationItem from './notification-item.component';

import { OtherUserType } from '../../redux/user/user.types';
import {
  selectCurrentUser,
  selectNotificationUsers,
} from '../../redux/user/user.selectors';
import { getOtherUserStart } from '../../redux/user/user.actions';

import { Notification } from '../../redux/follower/follower.types';
import {
  selectIsNotificationsDataLoading,
  selectNotifications,
  selectNotificationsQueryLength,
} from '../../redux/follower/follower.selectors';
import { getNotificationsStart } from '../../redux/follower/follower.actions';

import {
  selectNotificationUserMap,
  selectNotificationPostData,
  selectNotificationPostFiles,
} from '../../redux/post/post.selectors';
import {
  getSinglePostDataStart,
  getPostFileStart,
} from '../../redux/post/post.actions';
import { FileRequestType, UserType } from '../../redux/post/post.types';
import { useLazyLoading } from '../../hooks';

const NotificationsContainer: React.FC = () => {
  const [readyToRender, setReadyToRender] = useState(false);
  const [usersReady, setUsersReady] = useState(false);
  const [postFilesReady, setPostFilesReady] = useState(false);
  const [readyToFetchPhotos, setReadyToFetchPhotos] = useState(false);
  const currentUser = useSelector(selectCurrentUser);
  const notifications = useSelector(selectNotifications);
  const notificationUsers = useSelector(selectNotificationUsers);
  const notificationUserMap = useSelector(selectNotificationUserMap);
  const notificationPostData = useSelector(selectNotificationPostData);
  const notificationPostFiles = useSelector(selectNotificationPostFiles);
  const isNotificationsDataLoading = useSelector(
    selectIsNotificationsDataLoading
  );
  const notificationsQueryLength = useSelector(selectNotificationsQueryLength);
  const postDataFetchCount = useRef<Record<string, boolean>>({});
  const pageToFetch = useRef<number>(1);
  const dispatch = useDispatch();

  const { intersectionCounter, observedElementRef } = useLazyLoading(
    isNotificationsDataLoading,
    true,
    1000
  );

  let postsBucket: string;
  let profileBucket: string;

  if (process.env.NODE_ENV === 'production') {
    postsBucket = 'photo-share-app';
    profileBucket = 'photo-share-app-profile-photos';
  } else {
    postsBucket = 'photo-share-app-dev';
    profileBucket = 'photo-share-app-profile-photos-dev';
  }

  useEffect(() => {
    if (currentUser && !notifications) {
      dispatch(
        getNotificationsStart({
          userId: currentUser.id,
          pageToShow: 1,
          limit: 10,
        })
      );
    }
  }, [dispatch, currentUser, notifications]);

  useEffect(() => {
    const shouldFetchMoreNotifications = () =>
      notifications?.length &&
      notificationsQueryLength &&
      intersectionCounter !== pageToFetch.current &&
      notifications.length < notificationsQueryLength &&
      pageToFetch.current + 1 <= Math.ceil(notificationsQueryLength / 10);

    if (currentUser && shouldFetchMoreNotifications()) {
      dispatch(
        getNotificationsStart({
          userId: currentUser.id,
          limit: 10,
          pageToShow: ++pageToFetch.current,
        })
      );
    }
  }, [
    dispatch,
    currentUser,
    intersectionCounter,
    notifications,
    notificationsQueryLength,
  ]);

  useEffect(() => {
    if (notifications?.length) {
      notifications.forEach(({ fromUserId, postId }) => {
        if (!(notificationUsers && fromUserId in notificationUsers)) {
          dispatch(
            getOtherUserStart({
              type: OtherUserType.NOTIFICATION_USER,
              usernameOrId: fromUserId,
            })
          );
        }

        if (
          postId &&
          !(postId in postDataFetchCount.current) &&
          !notificationPostData.hasOwnProperty(postId)
        ) {
          postDataFetchCount.current[postId] = true;
          dispatch(getSinglePostDataStart({ postId, notificationPost: true }));
        }
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dispatch, notifications, notificationUsers]);

  useEffect(() => {
    if (!!notificationUsers) {
      const ready = notifications?.every(
        (notification: Notification) =>
          notificationUsers?.[notification.fromUserId]
      );

      setReadyToFetchPhotos(ready || false);
    }
  }, [notifications, notificationUsers]);

  useEffect(() => {
    if (
      Object.values(notificationPostData).length ===
        Object.values(postDataFetchCount.current).length &&
      Object.values(notificationPostData).length !==
        Object.values(notificationPostFiles).length
    ) {
      Object.values(notificationPostData).forEach((post) => {
        dispatch(
          getPostFileStart({
            s3Key: post.s3Key,
            bucket: postsBucket,
            user: UserType.notificationUser,
            fileRequestType: FileRequestType.notificationPost,
            isVideo: !!post.isVideo,
            ...(post.videoThumbnailS3Key && {
              videoThumbnailS3Key: post.videoThumbnailS3Key,
            }),
          })
        );
      });
    }
  }, [dispatch, postsBucket, notificationPostData, notificationPostFiles.size]);

  useEffect(() => {
    if (readyToFetchPhotos) {
      Object.values(notificationUsers).forEach((user) => {
        if (user.photo && !notificationUserMap.get(user.photo)) {
          dispatch(
            getPostFileStart({
              fileRequestType: FileRequestType.singlePost,
              bucket: profileBucket,
              s3Key: user.photo,
              user: UserType.notificationUser,
            })
          );
        }
      });

      setUsersReady(true);
    }
  }, [
    dispatch,
    profileBucket,
    readyToFetchPhotos,
    notificationUsers,
    notificationUserMap,
  ]);

  useEffect(() => {
    if (
      Object.values(notificationPostFiles).length ===
      Object.values(postDataFetchCount.current).length
    ) {
      setPostFilesReady(true);
    }
  }, [notificationPostFiles]);

  useEffect(() => {
    if (usersReady && postFilesReady) {
      setReadyToRender(true);
    }
  }, [usersReady, postFilesReady]);

  return (
    <Grid sx={{ display: 'flex', flexDirection: 'column', width: '100%' }}>
      {readyToRender &&
        notifications?.map((notification, idx) => {
          const user = notificationUsers?.[notification.fromUserId];
          const userPhotoInfo =
            (!!user.photo?.length && notificationUserMap.get(user.photo)) ||
            null;
          const postPhotoS3Key =
            notification.postId &&
            notificationPostData?.[notification.postId].s3Key;
          const postPhotoInfo =
            postPhotoS3Key && notificationPostFiles[postPhotoS3Key];

          return (
            <NotificationItem
              notification={notification}
              key={notification.id}
              user={user}
              userPhotoInfo={userPhotoInfo}
              postPhotoInfo={postPhotoInfo || null}
              custRef={
                idx === notifications.length - 1 ? observedElementRef : null
              }
            />
          );
        })}
      {!notifications?.length && (
        <Grid sx={{ display: 'flex', justifyContent: 'center', padding: 2 }}>
          <Typography>No notifications</Typography>
        </Grid>
      )}
    </Grid>
  );
};

export default NotificationsContainer;
