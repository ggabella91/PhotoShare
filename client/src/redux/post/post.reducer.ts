import { PostActions, PostActionTypes, PostState } from './post.types';

const INITIAL_STATE: PostState = {
  postData: null,
  postFiles: [],
  getPostDataError: null,
  getPostDataConfirm: null,
  getPostFileError: null,
  getPostFileConfirm: null,
  postError: null,
  postConfirm: null,
  profilePhotoFile: null,
  profilePhotoError: null,
  profilePhotoConfirm: null,
};

const postReducer = (state = INITIAL_STATE, action: PostActionTypes) => {
  switch (action.type) {
    case PostActions.UPDATE_PROFILE_PHOTO_SUCCESS:
      return {
        ...state,
        profilePhotoError: null,
        profilePhotoConfirm: 'Profile photo updated!',
      };
    case PostActions.CREATE_POST_SUCCESS:
      return {
        ...state,
        postError: null,
        postConfirm: 'Post uploaded!',
      };
    case PostActions.GET_POST_DATA_SUCCESS:
      return {
        ...state,
        postData: action.payload,
        getPostDataError: null,
        getPostDataConfirm: 'Post data fetched!',
      };
    case PostActions.GET_POST_FILE_SUCCESS:
      return {
        ...state,
        postFiles: [...state.postFiles, action.payload],
        getPostFileError: null,
        getPostFileConfirm: 'Post file fetched!',
      };
    case PostActions.GET_PROFILE_PHOTO_FILE_SUCCESS:
      return {
        ...state,
        profilePhotoFile: action.payload,
        getPostFileError: null,
        getPostFileConfirm: 'Profile photo file fetched!',
      };
    case PostActions.UPDATE_PROFILE_PHOTO_FAILURE:
      return {
        ...state,
        profilePhotoError: action.payload,
        profilePhotoConfirm: null,
      };
    case PostActions.CREATE_POST_FAILURE:
      return {
        ...state,
        postError: action.payload,
        postConfirm: null,
      };
    case PostActions.GET_POST_DATA_FAILURE:
      return {
        ...state,
        getPostDataError: action.payload,
        getPostDataConfirm: null,
      };
    case PostActions.GET_POST_FILE_FAILURE:
      return {
        ...state,
        getPostFileError: action.payload,
        getPostFileConfirm: null,
      };
    default:
      return state;
  }
};

export default postReducer;
