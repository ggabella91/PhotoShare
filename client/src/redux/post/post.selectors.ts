import { createSelector } from 'reselect';

import { AppState } from '../root-reducer';
import { PostState } from './post.types';

const selectPostState = (state: AppState): PostState => state.post;

export const selectPostData = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postData
);

export const selectPostDataFeedArray = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postDataFeedArray
);

export const selectPostReactionsArray = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postReactionsArray
);

export const selectPostReactionConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postReactionConfirm
);

export const selectPostReactionError = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postReactionError
);

export const selectGetPostDataError = createSelector(
  [selectPostState],
  (postState: PostState) => postState.getPostDataError
);

export const selectGetPostDataConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.getPostDataConfirm
);

export const selectGetPostFileError = createSelector(
  [selectPostState],
  (postState: PostState) => postState.getPostFileError
);

export const selectGetPostFileConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.getPostFileConfirm
);

export const selectGetPostReactionsError = createSelector(
  [selectPostState],
  (postState: PostState) => postState.getPostReactionsError
);

export const selectGetPostReactionsConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.getPostReactionsConfirm
);

export const selectPostFiles = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postFiles
);

export const selectPostError = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postError
);

export const selectPostConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postConfirm
);

export const selectProfilePhotoKey = createSelector(
  [selectPostState],
  (postState: PostState) => postState.profilePhotoKey
);

export const selectProfilePhotoFile = createSelector(
  [selectPostState],
  (postState: PostState) => postState.profilePhotoFile
);

export const selectOtherUserProfilePhotoFile = createSelector(
  [selectPostState],
  (postState: PostState) => postState.otherUserProfilePhotoFile
);

export const selectFollowPhotoFileArray = createSelector(
  [selectPostState],
  (postState: PostState) => postState.followPhotoFileArray
);

export const selectSuggestionPhotoFileArray = createSelector(
  [selectPostState],
  (postState: PostState) => postState.suggestionPhotoFileArray
);

export const selectReactorPhotoFileArray = createSelector(
  [selectPostState],
  (postState: PostState) => postState.reactorPhotoFileArray
);

export const selectUsersProfilePhotoConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.usersProfilePhotoConfirm
);

export const selectUpdateProfilePhotoError = createSelector(
  [selectPostState],
  (postState: PostState) => postState.profilePhotoError
);

export const selectUpdateProfilePhotoConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.profilePhotoConfirm
);

export const selectArchivePostError = createSelector(
  [selectPostState],
  (postState: PostState) => postState.archivePostError
);

export const selectArchivePostConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.archivePostConfirm
);

export const selectDeleteReactionConfirm = createSelector(
  [selectPostState],
  (postState: PostState) => postState.deleteReactionConfirm
);

export const selectDeleteReactionError = createSelector(
  [selectPostState],
  (postState: PostState) => postState.deleteReactionError
);

export const selectCommentToDelete = createSelector(
  [selectPostState],
  (postState: PostState) => postState.commentToDelete
);

export const selectShowCommentOptionsModal = createSelector(
  [selectPostState],
  (postState: PostState) => postState.showCommentOptionsModal
);

export const selectIsLoadingPostData = createSelector(
  [selectPostState],
  (postState: PostState) => postState.isLoadingPostData
);

export const selectPostMetaDataForUser = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postMetaDataForUser
);

export const selectPostLikingUsersArray = createSelector(
  [selectPostState],
  (postState: PostState) => postState.postLikingUsersArray
);
