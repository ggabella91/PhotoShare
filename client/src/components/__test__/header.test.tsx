import { shallow } from 'enzyme';
import React from 'react';
import { Header } from '../header/header.component';
import { PostFileReq } from '../../redux/post/post.types';
import { getPostFileStart } from '../../redux/post/post.actions';
import { signOutStart } from '../../redux/user/user.actions';

it('renders a header component', () => {
  // const fileReq: PostFileReq = { s3Key: 'key', bucket: 'bucket' };

  const headerWrapper = shallow(
    <Header
      currentUser={null}
      profilePhotoFile='photo'
      profilePhotoKey='photo-key'
      getPostFileStart={(fileReq) => getPostFileStart(fileReq)}
      signOutStart={() => signOutStart()}
    />
  );

  expect(headerWrapper).toMatchSnapshot();
});
