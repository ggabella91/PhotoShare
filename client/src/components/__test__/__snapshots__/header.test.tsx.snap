// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a header component 1`] = `
<div
  className="header"
>
  <NavLink
    className="app-name"
    to="/"
  >
    <h1
      className="title"
    >
      PhotoShare
    </h1>
  </NavLink>
</div>
`;
